client/transaction-pool/src/lib.rs

```diff
@@ -563,170 +585,320 @@ async fn prune_known_txs_for_block<Block: BlockT, Api: graph::ChainApi<Block = B
+			self.handle_enactment(hash, tree_route).await;
+		}
+
+		if let ChainEvent::Finalized { hash, tree_route } = event {
+			log::trace!(target:"txpool", "on-finalized enacted: {tree_route:?}, previously finalized: \
+						{prev_finalized_block:?}");
+			for hash in tree_route.iter().chain(std::iter::once(&hash)) {
+				if let Err(e) = self.pool.validated_pool().on_block_finalized(*hash).await {
+					log::warn!(
+						target: "txpool",
+						"Error [{}] occurred while attempting to notify watchers of finalization {}",
+						e, hash
+					)
 				}
-				.boxed()
-			},
+			}
 		}
 	}
 }
 
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactment was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Block::Hash,
+	recent_finalized_block: Block::Hash,
+}
+
+impl<Block> EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	fn new(recent_best_block: Block::Hash, recent_finalized_block: Block::Hash) -> Self {
+		EnactmentState { recent_best_block, recent_finalized_block }
+	}
+
+	/// computes tree route using provided ChainApi instance
+	fn tree_route<PoolApi: graph::ChainApi<Block = Block>>(
+		api: &PoolApi,
+		from: Block::Hash,
+		to: Block::Hash,
+	) -> Result<TreeRoute<Block>, String> {
+		match api.tree_route(from, to) {
+			Ok(tree_route) => Ok(tree_route),
+			Err(e) =>
+				return Err(format!(
+					"Error [{e}] occured while computing tree_route from {from:?} to \
+					 previously finalized: {to:?}"
+				)),
+		}
+	}
+
+	/// This function returns true and the tree_route if blocks enact/retract process (implemented
+	/// in handle_enactment function) should be performed based on:
+	/// - recent_finalized_block
+	/// - recent_best_block
+	/// - newly provided block in event
+	/// tree_route contains blocks to be enacted/retracted.
+	///
+	/// if enactment process is not needed (false, None) is returned
+	/// error message is returned when computing tree_route fails
+	fn update_and_check_if_new_enactment_is_valid<F>(
+		&mut self,
+		event: &ChainEvent<Block>,
+		tree_route: &F,
+	) -> Result<(bool, Option<TreeRoute<Block>>), String>
```

Comment:
```suggestion
	) -> Result<Option<TreeRoute<Block>>, String>
```
(If I'm not considering some future extension you already have in mind)
Currently on success function returns one of these two cases:
- `(false, None)`
- `(true, Some(tree_route))`

This can thus simplified by just returning:
- `None`
- `Some(tree_route)`

This requires then to change this routing consumers (i.e. true <=> tree_route.is_some())

Reply:
I've not finished to review the whole logic yet (it is still a bit of a moving target :-) )

Reply:
Hm... I would not mix the presence of `TreeRoute` with the boolean value indicating if process should proceed. The proper `TreeRoute` exists even if `None` is returned.
In my opinion having the clear `bool` value is more readable.

Reply:
[cf98315](https://github.com/paritytech/substrate/pull/12305/commits/cf983158c70ca80017360c67f9a45a9191ebebe5)

Reply:
My (suggested) approach is to keep simple things simple.
 Here, returning the route means the availability for that route to be used by the caller. There is a 1 to 1 mapping between the boolean val and the route presence in the Option.
Anyway, I see you've followed the suggestion 😀